### PR TITLE
CSV generator

### PR DESCRIPTION
**Pull request by masayag**
Generates resources for CSV installation and upgrades to be consumed by HCO.
In order to generate `manifests/vm-import-operator/v0.0.1/vm-import-operator.0.0.1.clusterserviceversion.yaml` run:

```
IMAGE_TAG=v0.0.1 QUAY_USER=kubevirt make gen-manifests
```

### CODE REVIEW

**Review by pkliczewski — COMMENTED** [review]
Looks good, do we want to update vm import crd status properties or will we do it only in the code?

**Review by pkliczewski — COMMENTED** [review]
Looks good to me. Feel free to remove WIP when ready.

**Review by pkliczewski — APPROVED** [review]

File: build/Dockerfile [hunk 1]
```diff
@@ -1,12 +1,23 @@
...
+RUN CGO_ENABLED=0 GOOS=linux go build -o /csv-generator github.com/kubevirt/vm-import-operator/tools/csv-generator
+
 FROM registry.access.redhat.com/ubi7/ubi-minimal:latest
 
 ENV OPERATOR=/usr/local/bin/vm-import-operator \
     USER_UID=1001 \
     USER_NAME=vm-import-operator
 
+ENV CSV_GENERATOR=/usr/bin/csv-generator \
+    USER_UID=1001 \
```
Comment (machacekondra) [inline build/Dockerfile]: Any reason to define USER_UID and USER_NAME twice? It's defined on line 11 & 12

File: pkg/operator/resources/operator/operator.go [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+					Type:      csvv1.InstallModeTypeAllNamespaces,
+					Supported: true,
+				},
+			},
+			InstallStrategy: csvv1.NamedInstallStrategy{
+				StrategyName:    "deployment",
+				StrategySpecRaw: json.RawMessage(strategySpecJSONBytes),
+			},
+			CustomResourceDefinitions: csvv1.CustomResourceDefinitions{
+				// TODO: should resource mappings crd be included here as well or only watched CRs ?
```
Comment (jakub-dzon) [inline pkg/operator/resources/operator/operator.go]: Is this TODO still valid?

File: hack/generate-manifests.sh [hunk 3]
```diff
@@ -28,7 +30,25 @@ for template in $templates; do
...
+	--image-pull-policy=${IMAGE_PULL_POLICY} \
+)
+if [[ ! -z "$rendered" ]]; then
+	echo -e "$rendered" > $file
+fi
+
+(cd ${PROJECT_ROOT}/tools/csv-generator && go clean)
+
 # TODO: Will be useful once examples are generated per version
+exit 0
```
Comment (machacekondra) [inline hack/generate-manifests.sh]: Why is here this exit? Maybe some testing leftover?

File: manifests/vm-import-operator/v0.0.1/vm-import-operator.0.0.1.clusterserviceversion.yaml [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: vm-import-operator.0.0.1
+  namespace: kubevirt-hyperconverged
+spec:
+  apiservicedefinitions: {}
+  customresourcedefinitions:
+    owned:
+    - description: Represents a virual machine import
```
Comment (machacekondra) [inline manifests/vm-import-operator/v0.0.1/vm-import-operator.0.0.1.clusterserviceversion.yaml]: s/virual/virtual

File: templates/vm-import-operator/VERSION/v2v_v1alpha1_virtualmachineimport_crd.yaml.in [hunk 5]
```diff
@@ -18,82 +18,256 @@ spec:
...
-              type: string
-            startVm:
-              description: 'StartVm defines if the VM should be started at the end of the import process'
-              type: boolean
+              required:
+              - name
+              type: object
             source:
-              description: 'VirtualMachineImportSourceSpec defines the definition of the source provider and mapping resources'
+              description: VirtualMachineImportSourceSpec defines the definition of
```
Comment (jakub-dzon) [inline templates/vm-import-operator/VERSION/v2v_v1alpha1_virtualmachineimport_crd.yaml.in]: Can "defines the definition" be rephrased?

File: templates/vm-import-operator/VERSION/v2v_v1alpha1_virtualmachineimport_crd.yaml.in [hunk 6]
```diff
@@ -18,82 +18,256 @@ spec:
...
+                the source provider and mapping resources
               properties:
                 ovirt:
-                  description: 'VirtualMachineImportOvirtSourceSpec defines the definition of the VM in oVirt'
+                  description: VirtualMachineImportOvirtSourceSpec defines the definition
```
Comment (jakub-dzon) [inline templates/vm-import-operator/VERSION/v2v_v1alpha1_virtualmachineimport_crd.yaml.in]: Can "defines the definition" be rephrased?

File: manifests/vm-import-operator/v0.0.1/vm-import-operator.0.0.1.clusterserviceversion.yaml [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+      displayName: Virual Machine import
```
Comment (machacekondra) [inline manifests/vm-import-operator/v0.0.1/vm-import-operator.0.0.1.clusterserviceversion.yaml]: Same here

File: templates/vm-import-operator/VERSION/v2v_v1alpha1_virtualmachineimport_crd.yaml.in [hunk 8]
```diff
@@ -18,82 +18,256 @@ spec:
...
-                        id:
-                          description: 'ID defines the ID of the VM in oVirt'
-                          type: string
-                        name:
-                          description: 'Name defines the name of the VM in oVirt'
-                          type: string
                         cluster:
-                          description: 'VirtualMachineImportOvirtSourceVMClusterSpec defines the identification of source VM cluster in oVirt'
+                          description: VirtualMachineImportOvirtSourceVMClusterSpec
+                            defines the definition of the source cluster of the VM
```
Comment (jakub-dzon) [inline templates/vm-import-operator/VERSION/v2v_v1alpha1_virtualmachineimport_crd.yaml.in]: Can "defines the definition" be rephrased?

**Review by jakub-dzon — CHANGES_REQUESTED** [review]

File: build/Dockerfile (hunk 1 above)
Comment (masayag) [reply build/Dockerfile]: nope, squashed to a single ENV statement

File: hack/generate-manifests.sh (hunk 3 above)
Comment (masayag) [reply hack/generate-manifests.sh]: the condition below never met. I'll remove the 'exit' and the dead code beneath

File: manifests/vm-import-operator/v0.0.1/vm-import-operator.0.0.1.clusterserviceversion.yaml (hunk 4 above)
Comment (masayag) [reply manifests/vm-import-operator/v0.0.1/vm-import-operator.0.0.1.clusterserviceversion.yaml]: interesting - seems the culprit is pkg/operator/resources/operator/operator.go in which we have:
```
 DisplayName: "Virual Machine import",
```
and as a result any generated file will carry that mistake. I'll fix the origin + generated resources.

File: pkg/operator/resources/operator/operator.go (hunk 2 above)
Comment (masayag) [reply pkg/operator/resources/operator/operator.go]: I'll remove the comment. we'll have to add generation of CRDs by code but they will be added differently and not in this context.

File: templates/vm-import-operator/VERSION/v2v_v1alpha1_virtualmachineimport_crd.yaml.in [hunk 9]
```diff
@@ -18,82 +18,254 @@ spec:
...
+                                type: string
+                            required:
+                            - source
+                            - target
+                            type: object
+                          type: array
+                      type: object
                     vm:
-                      description: 'VirtualMachineImportOvirtSourceVMSpec defines the identification of the VM in oVirt'
+                      description: VirtualMachineImportOvirtSourceVMSpec defines the
```
Comment (jakub-dzon) [inline templates/vm-import-operator/VERSION/v2v_v1alpha1_virtualmachineimport_crd.yaml.in:185]: One more "defines the definition"

**Review by jakub-dzon — CHANGES_REQUESTED** [review]

**Review by machacekondra — APPROVED** [review]

**Review by jakub-dzon — APPROVED** [review]

File: templates/vm-import-operator/VERSION/v2v_v1alpha1_virtualmachineimport_crd.yaml.in (hunk 9 above)
Comment (pkliczewski) [reply templates/vm-import-operator/VERSION/v2v_v1alpha1_virtualmachineimport_crd.yaml.in:185]: Done

**Pull request merged**